src/providers/ldap/sdap.c

```diff
@@ -409,92 +414,113 @@ int sdap_parse_entry(TALLOC_CTX *memctx,
+    objectClass_matched = !map;
+
+    do {
+        if (bvals) {
+            ber_memfree(bvals);
+            bvals = NULL;
+        }
+        lerrno = ldap_get_attribute_ber(sh->ldap, sm->msg, ber, &bv, &bvals);
+
+        /* Assignments before error-checking, as no de-ref involved... */
+        const char *attr_name = (char *)bv.bv_val;
+        const struct berval *setof_attr_val = bvals;
 
-    if (map) {
-        vals = ldap_get_values_len(sh->ldap, sm->msg, "objectClass");
-        if (!vals) {
+        if (lerrno != LDAP_SUCCESS) {
             DEBUG(SSSDBG_CRIT_FAILURE,
-                  "Unknown entry type, no objectClasses found!\n");
-            ret = EINVAL;
+                  "ldap_get_attribute_ber failed: %d(%s)\n",
+                  lerrno, sss_ldap_err2string(lerrno));
+            ret = EIO;
             goto done;
         }
 
-        for (i = 0; vals[i]; i++) {
-            if (objectclass_matched(map, vals[i]->bv_val, vals[i]->bv_len)) {
-                /* ok it's an entry of the right type */
-                break;
+        /* A NULL bv.bv_val w/ LDAP_SUCCESS indicates that "all data
+           has been consumed". */
+        if (attr_name == NULL)
+            break;
+
+        DEBUG(SSSDBG_TRACE_LIBS, "processing attribute [%s]\n", attr_name);
+
+        if (!objectClass_matched && strcmp(bv.bv_val, "objectClass") == 0) {
+            for (i = 0; setof_attr_val[i].bv_val; i++) {
+                if (objectclass_matched_ber(map, setof_attr_val[i])) {
+                    /* ok it's an entry of the right type */
+                    DEBUG(SSSDBG_TRACE_LIBS, "objectClass '%s' matched\n",
+                          setof_attr_val[i].bv_val);
+                    objectClass_matched = true;
+                    break;
+                }
+
+                DEBUG(SSSDBG_TRACE_LIBS, "objectClass '%s' did not match\n",
+                      setof_attr_val[i].bv_val);
             }
-        }
-        if (!vals[i]) {
-            DEBUG(SSSDBG_CRIT_FAILURE, "objectClass not matching: %s\n",
-                  map[0].name);
-            ldap_value_free_len(vals);
-            ret = EINVAL;
-            goto done;
-        }
-        ldap_value_free_len(vals);
-    }
 
-    str = ldap_first_attribute(sh->ldap, sm->msg, &ber);
-    if (!str) {
-        ldap_get_option(sh->ldap, LDAP_OPT_RESULT_CODE, &lerrno);
-        DEBUG(lerrno == LDAP_SUCCESS
-              ? SSSDBG_TRACE_LIBS
-              : SSSDBG_MINOR_FAILURE,
-              "Entry has no attributes [%d(%s)]!?\n",
-               lerrno, sss_ldap_err2string(lerrno));
-        if (map) {
-            ret = EINVAL;
-            goto done;
+            /* Strictly an optimization: don't process the objectClass
+               by iterating over the values again below. */
+            if (objectClass_matched)
+                goto skip;
+
+            /* A response is allowed to have only a single objectClass
```

Comment:
I don't understand this comment. Is this a particularity of Google's LDAP server? Does it allow several instances of any attribute except for `objectClass`? What does it do when `objectClass` is  multi-valued?